v2/client/invoker.go

```diff
@@ -39,50 +40,55 @@ type receiveInvoker struct {
 func (r *receiveInvoker) Invoke(ctx context.Context, m binding.Message, respFn protocol.ResponseFn) (err error) {
 	var isFinished bool
 	defer func() {
+		// skip finishing if we did it explicitly already
 		if !isFinished {
-			if err2 := m.Finish(err); err2 == nil {
-				err = err2
+			if fErr := m.Finish(err); fErr == nil {
+				err = nil
 			}
 		}
 	}()
 
-	e, err := binding.ToEvent(ctx, m)
-	if err != nil {
-		return err
-	}
+	var respMsg binding.Message
+	var result protocol.Result
+
+	e, eventErr := binding.ToEvent(ctx, m)
+	switch {
+	case eventErr != nil:
+		result = protocol.NewReceipt(false, "failed to convert Message to Event: %v", eventErr)
+
+	case e != nil && r.fn != nil:
+		var resp *event.Event
 
-	if e != nil && r.fn != nil {
-		resp, result := r.fn.invoke(ctx, *e)
+		resp, result = r.fn.invoke(ctx, *e)
+
+		if respFn == nil {
+			break
+		}
 
 		// Apply the defaulter chain to the outgoing event.
 		if resp != nil && len(r.eventDefaulterFns) > 0 {
 			for _, fn := range r.eventDefaulterFns {
 				*resp = fn(ctx, *resp)
 			}
 			// Validate the event conforms to the CloudEvents Spec.
-			if verr := resp.Validate(); verr != nil {
-				cecontext.LoggerFrom(ctx).Error(fmt.Errorf("cloudevent validation failed on response event: %v, %w", verr, err))
+			if vErr := resp.Validate(); vErr != nil {
+				cecontext.LoggerFrom(ctx).Error(fmt.Errorf("cloudevent validation failed on response event: %w", vErr))
 			}
 		}
 
-		// protocol can manual ack by the result
-		if respFn == nil {
-			if !protocol.IsACK(result) {
-				err = m.Finish(result)
-				isFinished = true
-			}
-			return
-		}
+		respMsg = (*binding.EventMessage)(resp)
+	}
 
-		var rm binding.Message
-		if resp != nil {
-			rm = (*binding.EventMessage)(resp)
+	if respFn == nil {
+		// let the protocol ACK based on the result
+		if !protocol.IsACK(result) {
```

Comment:
coud we not just return here and set `err = result`? let the deferred function call finish? because if IsAck, we will call finish anyway.

Reply:
That's what I wanted to do originally but I was not certain what was going on with that defer.

Done